File: solutions/python/algorithm/queue/Exercise4.py
from structure.dynamic.DynamicQueueADT import DynamicQueueADT

def elementos_en_comun(cola1: DynamicQueueADT, cola2: DynamicQueueADT) -> DynamicQueueADT:
    copia1 = copiar_cola(cola1)
    copia2 = copiar_cola(cola2)
    resultado = DynamicQueueADT()

    while not copia1.is_empty():
        actual = copia1.getElement()
        copia1.remove()

        temp = DynamicQueueADT()
        encontrado = False

        while not copia2.is_empty():
            valor = copia2.getElement()
            copia2.remove()
            if valor == actual and not encontrado:
                resultado.add(valor)
                encontrado = True
            temp.add(valor)

        copia2 = temp

    return resultado

def copiar_cola(original: DynamicQueueADT) -> DynamicQueueADT:
    copia = DynamicQueueADT()
    temp = DynamicQueueADT()

    while not original.is_empty():
        val = original.getElement()
        original.remove()
        copia.add(val)
        temp.add(val)

    while not temp.is_empty():
        original.add(temp.getElement())
        temp.remove()

    return copia


# Suponiendo que DynamicQueueADT tiene métodos: add, remove, getElement, is_empty

cola1 = DynamicQueueADT()
cola1.add(1)
cola1.add(2)
cola1.add(3)
cola1.add(4)

cola2 = DynamicQueueADT()
cola2.add(3)
cola2.add(4)
cola2.add(5)
cola2.add(6)

comun = elementos_en_comun(cola1, cola2)

print("Elementos en común:", end=" ")
while not comun.is_empty():
    print(comun.getElement(), end=" ")
    comun.remove()

File: solutions/javascript/src/algorithm/queue/Exercise4.js
import DynamicQueueADT from "../../structure/dynamic/DynamicQueueADT.js";

function copiarCola(original) {
  const copia = new DynamicQueueADT();
  const temp = [];

  while (!original.isEmpty()) {
    const val = original.getElement();
    original.remove();
    copia.add(val);
    temp.push(val);
  }

  for (const val of temp) {
    original.add(val);
  }

  return copia;
}

function elementosEnComun(cola1, cola2) {
  const copia1 = copiarCola(cola1);
  const copia2 = copiarCola(cola2);
  const resultado = new DynamicQueueADT();

  while (!copia1.isEmpty()) {
    const actual = copia1.getElement();
    copia1.remove();

    const temp = new DynamicQueueADT();
    let encontrado = false;

    while (!copia2.isEmpty()) {
      const val = copia2.getElement();
      copia2.remove();
      if (val === actual && !encontrado) {
        resultado.add(val);
        encontrado = true;
      }
      temp.add(val);
    }

    while (!temp.isEmpty()) {
      copia2.add(temp.getElement());
      temp.remove();
    }
  }

  return resultado;
}


// Suponiendo que DynamicQueueADT tiene métodos: add, remove, getElement, isEmpty

const cola1 = new DynamicQueueADT();
cola1.add(1);
cola1.add(2);
cola1.add(3);
cola1.add(4);

const cola2 = new DynamicQueueADT();
cola2.add(3);
cola2.add(4);
cola2.add(5);
cola2.add(6);

const comun = elementosEnComun(cola1, cola2);

let salida = "Elementos en común: ";
while (!comun.isEmpty()) {
  salida += comun.getElement() + " ";
  comun.remove();
}

console.log(salida);
